src/app/clusters/operational-state-server/operational-state-server.cpp

```diff
@@ -0,0 +1,417 @@
+}
+
+template <typename RequestT, typename FuncT>
+void OperationalStateServer::HandleCommand(HandlerContext & handlerContext, FuncT func)
+{
+    if (!handlerContext.mCommandHandled && (handlerContext.mRequestPath.mCommandId == RequestT::GetCommandId()))
+    {
+        RequestT requestPayload;
+
+        //
+        // If the command matches what the caller is looking for, let's mark this as being handled
+        // even if errors happen after this. This ensures that we don't execute any fall-back strategies
+        // to handle this command since at this point, the caller is taking responsibility for handling
+        // the command in its entirety, warts and all.
+        //
+        handlerContext.SetCommandHandled();
+
+        if (DataModel::Decode(handlerContext.mPayload, requestPayload) != CHIP_NO_ERROR)
+        {
+            handlerContext.mCommandHandler.AddStatus(handlerContext.mRequestPath,
+                                                     Protocols::InteractionModel::Status::InvalidCommand);
+            return;
+        }
+
+        func(handlerContext, requestPayload);
+    }
+}
+
+CHIP_ERROR OperationalStateServer::GetOperationalStateList(OperationalStateStructDynamicList ** operationalStateList, size_t & size)
+{
+    return mOperationalStateDataProvider.LoadOperationalStateList(endpointId, clusterId, operationalStateList, size);
+}
+
+void OperationalStateServer::ReleaseOperationalStateList(OperationalStateStructDynamicList * operationalStateList)
+{
+    mOperationalStateDataProvider.ReleaseOperationalStateList(operationalStateList);
+}
+
+CHIP_ERROR OperationalStateServer::ClearOperationalStateList()
+{
+    return mOperationalStateDataProvider.ClearOperationalStateList(endpointId, clusterId);
+}
+
+CHIP_ERROR OperationalStateServer::SetPhaseList(const PhaseList & phaseList)
+{
+    return mOperationalStateDataProvider.StorePhaseList(endpointId, clusterId, phaseList);
+}
+
+CHIP_ERROR OperationalStateServer::GetPhaseList(PhaseListCharSpan ** phaseList, size_t & size)
+{
+    return mOperationalStateDataProvider.LoadPhaseList(endpointId, clusterId, phaseList, size);
+}
+
+void OperationalStateServer::ReleasePhaseList(PhaseListCharSpan * phaseList)
+{
+    mOperationalStateDataProvider.ReleasePhaseList(phaseList);
+}
+
+CHIP_ERROR OperationalStateServer::ClearPhaseStateList()
+{
+    return mOperationalStateDataProvider.ClearPhaseStateList(endpointId, clusterId);
+}
+
+CHIP_ERROR OperationalStateServer::SetOperationalState(OperationalStateStruct & opState)
+{
+    OperationalStateStructDynamicList op;
+
+    op.operationalStateID = static_cast<Clusters::OperationalState::OperationalStateEnum>(opState.OperationalStateID);
+
+    ReturnErrorOnFailure(mOperationalStateDataProvider.UseOpStateIDGetOpStateStruct(endpointId, clusterId, op));
+
+    if (&operationalState != &opState)
+    {
+        char * dest = const_cast<char *>(operationalState.OperationalStateLabel);
```

Comment:
If you have to `const_cast`, chances are you have a bug....  I am pretty sure there should not be any const_cast here.